packages/flutter/lib/src/widgets/scrollable_repaint_boundary.dart

```diff
@@ -0,0 +1,148 @@
+      boundary.value = true;
+    } else if (notification is ScrollUpdateNotification) {
+      boundary.value = true;
+    } else if (notification is ScrollEndNotification) {
+      boundary.value = false;
+    }
+    return false;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return NotificationListener<ScrollNotification>(
+      onNotification: _onScrollUpdateNotification,
+      child: _ScrollableRepaintBoundaryNotifier(
+        state: boundary,
+        child: widget.child,
+      ),
+    );
+  }
+}
+
+/// An inherited widget that listens to the scroll state and notifies the
+/// child [ScrollableRepaintBoundary] to enable/disable the repaint
+/// boundary.
+class _ScrollableRepaintBoundaryNotifier extends InheritedWidget {
+  const _ScrollableRepaintBoundaryNotifier({
+    required super.child,
+    required this.state,
+  });
+
+  final ValueNotifier<bool> state;
+
+  @override
+  bool updateShouldNotify(covariant _ScrollableRepaintBoundaryNotifier oldWidget) {
+    return oldWidget.state != state;
+  }
+}
+
+/// A [RepaintBoundary] specialized for the children scrollable widgets.
+///
+/// Generally speaking, dividing a scene with repaint boundaries is only
+/// advantageous if it would both stabilize the picture and only to the point
+/// where there are a few (1-3) dozen pictures in a frame. If a picture is already
+/// protected by some parent repaint boundary and does not change, further dividing
+/// it with more repaint boundaries is counter productive as it degrades the
+/// efficiency of the Flutter engine's raster cache.
+///
+/// In the case of scrolling widgets, dividing each child item with a repaint boundary
+/// is usually ideal as it both prevents them from changing during the scroll
+/// (allowing them to be cached) and protects siblings from hover or material ink
+/// splash animations.
+///
+/// Nevertheless, there are several scenarios where this is counter-productive. If
+/// the list is nested in another list, the nested repaint boundaries will often
+/// create too many pictures for the raster cache to be effective. These pictures are
+/// often so small as to not meet the threshold for cache admission. If this list is
+/// shrink-wrapped or on a cross-axis but not able to scroll, then the most common
+/// case that makes repaint boundaries effective is removed.
+///
+/// Another scenario is where a list is used instead of a Row/Column/Wrap because
+/// its children may overflow, but often don't. In this case, the benefits observed
+/// during scrolling are also removed.
+///
+/// This widget attempts to address these cases, by making the repaint boundary used
+/// to wrap scrollable children conditional on whether the scrollable has ever
+/// scrolled before. This solution is only a heuristic, and could be adjusted to take
+/// into account whether or not the child has repainted or whether the scrollable
+/// is likely to scroll again.
+class ScrollableRepaintBoundary extends StatefulWidget {
```

Comment:
Actually, I think you can omit this and the associated state object entirely and only have the _ConditionalRepaintBoundary. It can establish and maintain a dependency in `createRenderObject`/`updateRenderObject`.